xarray/tests/test_combine.py

```diff
@@ -686,6 +716,14 @@ def test_combine_by_coords(self):
     def test_empty_input(self):
         assert_identical(Dataset(), combine_by_coords([]))
 
+    def test_combine_coords_mixed_datasets_arrays(self):
+        objs = [
+            DataArray([0, 1], dims=("x"), coords=({"x": [0, 1]})),
+            Dataset({"x": [2, 3]}),
+        ]
+        with pytest.raises(ValueError, match="without providing an explicit name"):
+            combine_by_coords(objs)
```

Comment:
Do we need a similar test for `combine_nested`? Even if it passes through fine?

Reply:
I could add a similar test for `combine_nested`. Right now `combine_nested` will raise `AttributeError` when run with the test above since the DataArray doesn't have a name.

Reply:
We should crystallise what we expect the behaviour to be in the form of a test. If the `AttributeError` currently returned is informative, then that's fine, if it's not then we should consider changing something to raise a better message.

Reply:
I decided that the `AttributeError` currently returned is rather confusing. I would rather declare up front in the combine* functions that such a mixed input (both unnamed `DataArray`s and `Dataset`s) is not allowed. This way a user of these functions can pretty quickly see that they are passing in an invalid set of inputs. Both functions now return `ValueError` along with an informative message.